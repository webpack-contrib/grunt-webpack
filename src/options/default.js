"use strict";

const mergeWith = require("lodash/mergeWith");

const gruntOptions = {
  failOnError: (options) => {
    // if watch enabled default to failOnError false
    return Array.isArray(options)
      ? options.every((option) => !option.watch)
      : !options.watch;
  },
  progress: process.stdout.isTTY,
  storeStatsTo: null,
  keepalive: (options) => {
    // if watch enabled default to keepalive true
    return Array.isArray(options)
      ? options.some((option) => option.watch)
      : !!options.watch;
  },
  watch: null,
};

const webpackOptions = {
  stats: {
    cached: false,
    cachedAssets: false,
    colors: true,
  },
  cache: (options) => {
    // if watch enabled also default to cache true
    return Array.isArray(options)
      ? options.some((option) => option.watch)
      : !!options.watch;
  },
};

const webpackDevServerOptions = {
  devServer: {
    host: "localhost",
  },
  stats: {
    cached: false,
    cachedAssets: false,
    colors: true,
  },
};

// eslint-disable-next-line consistent-return
function mergeCustomize(a, b) {
  if (Array.isArray(a) && Array.isArray(b)) {
    return a.concat(b);
  }
}

function mergeOptions(defaultOptions, options, targetOptions) {
  if (Array.isArray(targetOptions) && Array.isArray(options)) {
    if (targetOptions.length !== options.length) {
      throw new Error(
        "Cannot have both `options` and `target` be an array with different length. " +
          "If using arrays for both please ensure they are the same size.",
      );
    }
    return targetOptions.map((opt, index) =>
      mergeWith({}, defaultOptions, options[index], opt, mergeCustomize),
    );
  }

  if (Array.isArray(targetOptions)) {
    return targetOptions.map((opt) =>
      mergeWith({}, defaultOptions, options, opt, mergeCustomize),
    );
  } else if (Array.isArray(options)) {
    return options.map((opt) =>
      mergeWith({}, defaultOptions, opt, targetOptions, mergeCustomize),
    );
  }

  return mergeWith({}, defaultOptions, options, targetOptions, mergeCustomize);
}

exports.gruntOptions = gruntOptions;
exports.webpackOptions = webpackOptions;
exports.webpackDevServerOptions = webpackDevServerOptions;

exports.mergeOptions = mergeOptions;
