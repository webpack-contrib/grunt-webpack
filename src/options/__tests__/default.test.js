import { gruntOptions, mergeOptions } from "../default";

describe("default options", () => {
  describe("keepalive", () => {
    test("keepalive without watch", () => {
      expect(gruntOptions.keepalive({})).toBe(false);
    });

    test("keepalive with watch true", () => {
      expect(gruntOptions.keepalive({ watch: true })).toBe(true);
    });

    test("keepalive with watch false", () => {
      expect(gruntOptions.keepalive({ watch: false })).toBe(false);
    });

    test("keepalive without watch in array", () => {
      expect(gruntOptions.keepalive([{}])).toBe(false);
    });

    test("keepalive with watch true in array", () => {
      expect(gruntOptions.keepalive([{ watch: true }])).toBe(true);
    });

    test("keepalive with watch false in array", () => {
      expect(gruntOptions.keepalive([{ watch: false }])).toBe(false);
    });
  });

  describe("failOnError", () => {
    test("failOnError without watch", () => {
      expect(gruntOptions.failOnError({})).toBe(true);
    });

    test("failOnError with watch true", () => {
      expect(gruntOptions.failOnError({ watch: true })).toBe(false);
    });

    test("failOnError with watch false", () => {
      expect(gruntOptions.failOnError({ watch: false })).toBe(true);
    });

    test("failOnError without watch in array", () => {
      expect(gruntOptions.failOnError([{}])).toBe(true);
    });

    test("failOnError with watch true in array", () => {
      expect(gruntOptions.failOnError([{ watch: true }, {}])).toBe(false);
    });

    test("failOnError with watch false in array", () => {
      expect(gruntOptions.failOnError([{ watch: false }, {}, {}])).toBe(true);
    });
  });

  describe("mergeOptions", () => {
    test("all single objects", () => {
      expect(
        mergeOptions({ default: 1 }, { options: 1 }, { target: 1 }),
      ).toMatchSnapshot();
    });

    test("merges sub arrays", () => {
      expect(
        mergeOptions({ array: [1] }, { array: [2] }, { array: [3] }),
      ).toMatchSnapshot();
    });

    test("options as array", () => {
      expect(
        mergeOptions({ default: 1 }, [{ options1: 1 }, { options2: 1 }], {
          target: 1,
        }),
      ).toMatchSnapshot();
    });

    test("target as array", () => {
      expect(
        mergeOptions(
          { default: 1 },
          {
            options: 1,
          },
          [{ target1: 1 }, { target2: 1 }],
        ),
      ).toMatchSnapshot();
    });

    test("options and target as array", () => {
      expect(
        mergeOptions(
          { default: 1 },
          [{ options1: 1 }, { options2: 1 }],
          [{ target1: 1 }, { target2: 1 }],
        ),
      ).toMatchSnapshot();
    });

    test("fails on options and target as unequal array", () => {
      expect(() =>
        mergeOptions(
          { default: 1 },
          [{ options1: 1 }, { options2: 1 }, { options3: 1 }],
          [{ target1: 1 }, { target2: 1 }],
        ),
      ).toThrowError();
    });
  });
});
