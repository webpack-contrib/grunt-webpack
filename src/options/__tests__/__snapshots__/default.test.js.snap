// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default options mergeOptions all single objects 1`] = `
Object {
  "default": 1,
  "options": 1,
  "target": 1,
}
`;

exports[`default options mergeOptions merges sub arrays 1`] = `
Object {
  "array": Array [
    1,
    2,
    3,
  ],
}
`;

exports[`default options mergeOptions options and target as array 1`] = `
Array [
  Object {
    "default": 1,
    "options1": 1,
    "target1": 1,
  },
  Object {
    "default": 1,
    "options2": 1,
    "target2": 1,
  },
]
`;

exports[`default options mergeOptions options as array 1`] = `
Array [
  Object {
    "default": 1,
    "options1": 1,
    "target": 1,
  },
  Object {
    "default": 1,
    "options2": 1,
    "target": 1,
  },
]
`;

exports[`default options mergeOptions target as array 1`] = `
Array [
  Object {
    "default": 1,
    "options": 1,
    "target1": 1,
  },
  Object {
    "default": 1,
    "options": 1,
    "target2": 1,
  },
]
`;
